File: app/MMarket_old/c/src/server/settings/calendar.c
// Copyright 16-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "server/settings/calendar.h"
#include "dmc/cgi.h"
#include "data/MarketDay.h"
#include "db/global.h"
#include "db/calendar.h"

char *calendar_process (AsyncActor *ac, Map *mrq) {
  CGI_GET_STR(rq, mrq);
  if (str_eq(rq, "idata")) {
    // Map[Js]
    Map *rp = map_new();
    map_put(rp, "timeStamp", js_ws(global_time_stamp()));
    map_put(rp, "general", timetable_to_js(calendar_general()));
    map_put(rp, "holidays", arr_to_js(calendar_holidays(), (FTO)js_ws));
    map_put(
      rp,
      "specialDays",
      arr_to_js(calendar_special_days(), (FTO)marketDay_to_js)
    );
    return cgi_ok(rp);
  };
  if (str_eq(rq, "setGeneral")) {
    CGI_GET_STR(timeStamp, mrq);
    CGI_GET(Timetable *, timetable, timetable_from_js, mrq);

    // Map[Js]
    Map *rp = map_new();
    if (global_check_time_stamp(timeStamp)) {
      char *tmst;
      void fn () {
        calendar_set_general(timetable);
        tmst = global_set_time_stamp();
      }
      asyncActor_wait(ac, fn);
      map_put(rp, "timeStamp", js_ws(tmst));
      map_put(rp, "ok", js_wb(1));
    } else {
      map_put(rp, "ok", js_wb(0));
    }
    return cgi_ok(rp);
  }

  if (str_eq(rq, "setSpecialDays")) {
    CGI_GET_STR(timeStamp, mrq);
    CGI_GET_ARR(specialDays, mrq);

    // Map[Js]
    Map *rp = map_new();
    if (global_check_time_stamp(timeStamp)) {
      char *tmst;
      void fn () {
        calendar_set_special_days(
          arr_map(specialDays, (FCOPY)marketDay_from_js)
        );
        tmst = global_set_time_stamp();
      }
      asyncActor_wait(ac, fn);
      map_put(rp, "timeStamp", js_ws(tmst));
      map_put(rp, "ok", js_wb(1));
    } else {
      map_put(rp, "ok", js_wb(0));
    }
    return cgi_ok(rp);
  }

  EXC_ILLEGAL_ARGUMENT(
    "rq",
    "idata | setGeneral | setHolidays | setSpecialDays",
    rq
  )
  return NULL; // Unreachable
}

File: app/KtMarket/go/server/pgs/settings/calendar/calendar.go
// Copyright 05-Jun-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Settings-Calendar page
package calendar

import (
	"github.com/dedeme/KtMarket/data/calendar"
	"github.com/dedeme/KtMarket/db"
	"github.com/dedeme/ktlib/arr"
	"github.com/dedeme/ktlib/cgi"
	"github.com/dedeme/ktlib/js"
	"github.com/dedeme/ktlib/thread"
)

func Process(ck string, mrq cgi.T) string {
	rq := js.Rs(mrq["rq"])
	switch rq {
	case "idata":
		var cl string
		thread.Sync(func() {
			cl = db.CalendarTb().ReadJs()
		})
		return cgi.Rp(ck, cgi.T{
			"calendar": cl,
		})
	case "setGeneral":
		thread.Sync(func() {
			cl := db.CalendarTb().Read()
			cl.General = calendar.TimetableFromJs(mrq["timetable"])
			db.CalendarTb().Write(cl)
		})
		return cgi.RpEmpty(ck)
	case "setHolidays":
		thread.Sync(func() {
			cl := db.CalendarTb().Read()
			cl.Holidays = arr.Map(js.Ra(mrq["holidays"]), func(j string) string {
				return js.Rs(j)
			})
			db.CalendarTb().Write(cl)
		})
		return cgi.RpEmpty(ck)
	case "setSpecialDays":
		thread.Sync(func() {
			cl := db.CalendarTb().Read()
			cl.SpecialDays = arr.Map(
				js.Ra(mrq["specialDays"]),
				func(j string) *calendar.MarketDayT {
					return calendar.MarketDayFromJs(j)
				},
			)
			db.CalendarTb().Write(cl)
		})
		return cgi.RpEmpty(ck)
	default:
		panic("Value of rq (" + rq + ") is not valid")
	}
}
